### PR TITLE
Add txscript pay_to_address_script and extract_script_pub_key_address

### PR DESCRIPTION
**Pull request by tiram88**
Enables conversion from `Address `to `ScriptPublicKey` and backwards.

Some notable changes/enhancements:
- `Address` is enhanced thanks to a `Version` enum making the 3 supported kinds of addresses explicit.
- `Address` payload is turned into a `SmallVec` instead of a `Vec`.
- `ScriptClass` is added to txscript.

### CODE REVIEW

File: crypto/txscript/src/script_class.rs [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub enum Error {
+    InvalidScriptClass(String),
+}
+
+impl Display for Error {
```
[the remark this answers is not in the document]
Comment (tiram88) [reply crypto/txscript/src/script_class.rs:14]: ok, so I refactor tmrlvi's code

File: consensus/core/src/testutils/op_true_script.rs [hunk 2]
```diff
@@ -11,7 +11,7 @@ pub fn op_true_script() -> (ScriptPublicKey, Vec<u8>) {
     // TODO: use txscript.OpTrue instead when available
     let redeem_script = vec![81u8];
 
-    // TODO: use txscript.PayToScriptHashScript(redeemScript) when available
+    // TODO: use txscript::pay_to_script_hash_script(redeemScript)
```
[the remark this answers is not in the document]
Comment (tiram88) [reply consensus/core/src/testutils/op_true_script.rs:14]: Doing so is creating a circular dependency between consensus-core and txscript.
So I move testutils altogether to `txscript::test_helpers`.

File: crypto/txscript/src/script_class.rs [hunk 3]
```diff
@@ -44,13 +44,17 @@ const PUB_KEY_ECDSA: &str = "pubkeyecdsa";
...
 
 impl ScriptClass {
-    pub fn from_script(script_public_key: &[u8]) -> Self {
-        if Self::is_pay_to_pubkey(script_public_key) {
-            ScriptClass::PubKey
-        } else if Self::is_pay_to_pubkey_ecdsa(script_public_key) {
-            Self::PubKeyECDSA
-        } else if Self::is_pay_to_script_hash(script_public_key) {
-            Self::ScriptHash
+    pub fn from_script(script_public_key: &[u8], version: ScriptPublicKeyVersion) -> Self {
```
Comment (someone235) [inline crypto/txscript/src/script_class.rs:47]: Maybe it should just get a ScriptPublicKey and separate it internally to script and version?

File: crypto/txscript/src/standard.rs [hunk 4]
```diff
@@ -74,6 +73,55 @@ pub fn extract_script_pub_key_address(script_public_key: &ScriptPublicKey, prefi
...
+        let signature_script = pay_to_script_hash_signature_script(redeem_script, vec![]);
+
+        let previous_outpoint = TransactionOutpoint::new(tx_to_spend.id(), 0);
+        let input = TransactionInput::new(previous_outpoint, signature_script, MAX_TX_IN_SEQUENCE_NUM, 1);
+        let output = TransactionOutput::new(tx_to_spend.outputs[0].value - fee, script_public_key);
+        Transaction::new(TX_VERSION, vec![input], vec![output], 0, SUBNETWORK_ID_NATIVE, 0, vec![])
+    }
+
+    pub fn pay_to_script_hash_signature_script(redeem_script: Vec<u8>, signature: Vec<u8>) -> Vec<u8> {
+        // This is just a hack until we have ScriptBuilder
```
Comment (someone235) [inline crypto/txscript/src/standard.rs:112]: Add TODO so we can find it later

**Review by someone235 — CHANGES_REQUESTED** [review]

File: crypto/addresses/src/lib.rs [hunk 5]
```diff
@@ -128,14 +122,16 @@ impl TryFrom<u8> for Version {
 }
 
 /// Size of the payload vector of an address.
+///
+/// This size is the smallest SmallVec size greater or equal to the largest
+/// possible payload, which is 33 for [`Version::PubKeyECDSA`].
 pub const PAYLOAD_VECTOR_SIZE: usize = 36;
```
Comment (someone235) [inline crypto/addresses/src/lib.rs:128]: So why 36?

**Review by someone235 — CHANGES_REQUESTED** [review]

File: crypto/txscript/src/script_class.rs (hunk 3 above)
Comment (tiram88) [reply crypto/txscript/src/script_class.rs:47]: given the use in go, makes sense, ok

**Review by someone235 — APPROVED** [review]

**Pull request merged**